### PR TITLE
To moments

### PR DESCRIPTION
**Pull request by apragsdale**
Getting this up here, though not quite finished. I think I've got it working with pure split/branch demographies, but haven't implemented merger/admixture/pulse events yet.

One thing that is confusing me, that you might be able to help with @grahamgower, is I've tried to alter to `__init__.py` in the `demes/convert` directory to import the moments conversion methods. But I get import errors (`ModuleNotFoundError: No module named 'demes.convert.moments_'`). Do I also need to change something somewhere else, maybe in the setup.cfg or similar? It's been stumping me...

### CODE REVIEW

**Comment by grahamgower** [comment]
Awesome!

> One thing that is confusing me, that you might be able to help with @grahamgower, is I've tried to alter to `__init__.py` in the `demes/convert` directory to import the moments conversion methods. But I get import errors (`ModuleNotFoundError: No module named 'demes.convert.moments_'`). Do I also need to change something somewhere else, maybe in the setup.cfg or similar? It's been stumping me...

I pulled down your changes, and uncommented the import in `demes/convert/__init__.py` and it works for me. What were you doing exactly?
```Python
$ pwd
/home/grg/src/demes
$ python
Python 3.8.6 (default, Sep 30 2020, 04:00:38)
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import demes
>>> import demes.convert
>>> demes.convert.SFS
<function SFS at 0x7fd00eb569d0>
>>> demes.__file__
'/home/grg/src/demes/demes/__init__.py'
>>> import moments
>>> moments.__file__
'/home/grg/src/moments/moments/__init__.py'
>>>
>>> import sys
>>> sys.path
['', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '/home/grg/.local/lib/python3.8/site-packages', '/home/grg/src/demography', '/home/grg/src/moments', '/home/grg/src/dadi', '/home/grg/src/kastore/python', '/home/grg/src/genomatnn', '/home/grg/.local/lib/python3.8/site-packages/tensorflow-2.3.0-py3.8-linux-x86_64.egg', '/home/grg/src/stdpopsim', '/home/grg/.local/lib/python3.8/site-packages/tskit-0.3.0-py3.8-linux-x86_64.egg', '/home/grg/src/keras-vis', '/home/grg/src/demes', '/usr/lib/python3.8/site-packages']
```

What this needs to work, is for the user to have installed both `demes` and `moments` packages so they can be `import`ed. I personally do that in a lazy way so I can make changes to the sources and have those changes reflected when I do the import---I installed `moments` and `demes` using `python setup.py develop --user` from their respective git-cloned folders. This is probably atypical and not recommended in general (the recommended approach is to use virtual environments, but I plain don't like 'em).

**Comment by grahamgower** [comment]
And regards the `mypy` CI errors, you need to add a couple of lines to`setup.cfg` to tell mypy that moments doesn't have type annotations. There are currently four other examples of this workaround at the bottom of setup.cfg, for strictyaml, numpy, msprime, and stdpopsim.

**Comment by grahamgower** [comment]
I haven't yet gone through this in detail, but do you think we could unconditionally import `moments` and have your shiny new tests running in CI?

What I was thinking, is that `import demes` would require minimal dependencies (those listed in the `setup.cfg`). All the extra conversion stuff will live in a separately imported namespace `demes.convert`. If we only `import demes.convert` in the test suite, then we just need additional dependencies for developers and CI (currently listed in `requirements.txt`... maybe that should be renamed). Of course, I'm open to other ways of structuring this.

**Comment by molpopgen** [comment]
Is all the effort worth it, given that we'd undo it all later?  If the long term plan is to not have the conversion code here (and I agree with that plan), then setting up the CI doesn't seem worthwhile.

**Comment by grahamgower** [comment]
I'm expecting that we'll make some changes to the core parts of `demes` based on what we learn while implementing correct conversion functions. So the tests will ensure that we have consistency with the conversion functions, and can meet our goal(s) of having correctness with a flexible API. Yes, the conversion functions will move to other repositories, but it will be less work to do that if we have passing tests already.

**Comment by molpopgen** [comment]
Sounds good.  I can also imagine having CI include these other tools and running tests to guard against future regressions.

**Comment by grahamgower** [comment]
> I can also imagine having CI include these other tools and running tests to guard against future regressions.

Yeah, absolutely, that sounds like a fine plan!

**Comment by grahamgower** [comment]
@apragsdale it should be sufficient to add a line like this to requirements.txt in order to get moments installed for CI.
```moments @ git+https://bitbucket.org/simongravel/moments.git```

But I couldn't get moments installed via pip, so this fails. I opened an issue for this over on bitbucket. https://bitbucket.org/simongravel/moments/issues/36/cannot-install-moments-using-pip-install

**Comment by molpopgen** [comment]
I've used that method to install it on my own machines and it has worked.  What is the error that you are getting?

**Comment by molpopgen** [comment]
I see the other now in your issue report.  I've never seen that before, but I also haven't had to install it for several months. Maybe something has changed?

**Comment by apragsdale** [comment]
I'm getting the moments setup.py fixed up today, as Graham suggested. So hopefully once that is resolved, that won't be a hangup for any of this.

**Comment by apragsdale** [comment]
@grahamgower now that `moments` installation is a lot smoother, what exactly did you have in mind for importing msprime, moments, etc for the tests? Is it as simple as placing `moments @ git+https://bitbucket.org/simongravel/moments.git` in the `requirements.txt`? I think I'm getting close to have this wrapped up - just a few corner cases to finish up.

**Comment by grahamgower** [comment]
Yeah, exactly. I think you can even specify the version tag (I'm not sure of the syntax for that though, so might need some googling+trial/error). Then, if you remove the import protection from the test file, that will ensure that the tests really do get run during CI.

File: requirements.txt [hunk 1]
```diff
@@ -8,3 +8,4 @@ sphinx==3.2.1
 sphinx_rtd_theme==0.5.0
 stdpopsim==0.1.2
 strictyaml==1.1.0
+git+git://bitbucket.org/simongravel/moments@moments-1.0.5#egg=moments
```
Comment (grahamgower) [inline requirements.txt:11]: I reckon you'll need to use a `git+https://` url here.

**Comment by apragsdale** [comment]
Ah ok - so the install requires both `numpy` and `cython` to be installed to be able to run the `setup.py`... So in a completely fresh environment, I'm not sure how to ensure that those are installed before trying to install moments.

**Comment by grahamgower** [comment]
> Ah ok - so the install requires both `numpy` and `cython` to be installed to be able to run the `setup.py`... So in a completely fresh environment, I'm not sure how to ensure that those are installed before trying to install moments.

Supposedly, this chicken-and-egg problem is solved via the use of a suitably configured `pyproject.toml` (to be placed in the moments repo). I've never tried it, but there's an example using cython and numpy halfway down [this page](https://levelup.gitconnected.com/how-to-deploy-a-cython-package-to-pypi-8217a6581f09). And then we need to live in a magical wonderland where everyone has an up-to-date pip that recognises `pyproject.toml` (but at least we can upgrade that in the CI if it's too old).

Python packaging sucks!

**Comment by apragsdale** [comment]
Ok, I'll give that a try - might require a bumped moments version in that case. Thanks for the tips!

**Comment by apragsdale** [comment]
Yesssss - I'm very happy that looks to have worked properly. Thanks so much @grahamgower. It also just improves the `moments` install and setup overall. This has been a great learning experience, even though python packing is a PITA!

**Comment by apragsdale** [comment]
Ok - pretty happy where this is at. I'll wait to merge until you have a chance to make sure it all looks ok, especially the moments install and import.

**Review by grahamgower — APPROVED** [review]
Looks good @apragsdale. There's a bunch of stuff here that's over my head, but it occurs to me that there might be some functions here that will be useful more generally? Having a `prune(deme_graph, samples)` function seems like something we could consider?

File: demes/convert/moments_.py [hunk 2]
```diff
@@ -0,0 +1,875 @@
...
+# so the demography cannot have more than five populations at any given time.
+
+
+def SFS(g, sampled_demes, sample_sizes, sample_times=None, Ne=None, unsampled_n=4):
+    """
+    Takes a deme graph and computes the site frequency spectrum using ``moments``.
+
+    :param g: A demes graph from which to compute the SFS.
+    :param sampled_demes: A list of demes to take samples from. We can repeat
+        demes, as long as the sampling occurs at distinct times.
```
[the remark this answers is not in the document]
Comment (apragsdale) [reply demes/convert/moments_.py:19]: Yes, very good catch. They are deme IDs.
